### PR TITLE
CRIMRE-193 only auth extant users

### PR DESCRIPTION
**Pull request by timpeat**
## Description of change

Only authenticate users if they exist in the review database. 
Active users are authenticated if they are found by auth_oid.
Users pending activation are authenticated if they are found by email address.
Pending activation users are activated on authentication.
Once activated a user cannot be found by email address (in the context of authentication).

## Link to relevant ticket
[CRIMRE-193](https://dsdmoj.atlassian.net/browse/CRIMRE-193)

## Notes for reviewer

Be good to find out what sort of logging/monitoring is expected around authentication. I'd like to move User.authenticate! into a service or an authentication aggregate.

## How to manually test the feature

[CRIMRE-193]: https://dsdmoj.atlassian.net/browse/CRIMRE-193?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -20,5 +20,43 @@ def name_for(id)
         I18n.t('values.deleted_user_name')
       end
     end
+
+    #
+    # TODO find a better home.
+    #
```
Comment (willmcb) [inline app/models/user.rb:26]: The common term is "authenticate" so I guess that tells us something! :) 

We could have an authentication service etc.

File: lib/warden/strategies/azure_ad.rb [hunk 2]
```diff
@@ -2,27 +2,14 @@ module Warden
...
     class AzureAd < Warden::Strategies::Base
       def authenticate!
-        omniauth_user = env['omniauth.auth']
+        info = env.dig('omniauth.auth', 'info')
 
-        if omniauth_user
-          info = env['omniauth.auth']['info']
-
-          success! find_and_update_user(info)
+        if info && (user = User.authenticate!(info))
```
Comment (willmcb) [inline lib/warden/strategies/azure_ad.rb:7]: Perhaps move the instantiation of user out of the statement here

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -23,4 +23,99 @@
...
+        allow(described_class).to receive(:find_by).with(
+          { auth_oid: auth_info['auth_oid'] }
+        ).and_return user
+      end
+
+      it 'returns the active user' do
+        expect(authenticate!).to be user
+      end
+
+      it "sets the user's name, email and last_auth_at" do
```
Comment (willmcb) [inline spec/models/user_spec.rb:65]: single quotes for consistancy 👍

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -23,4 +23,99 @@
...
+        allow(described_class).to receive(:find_by)
+          .with({ email: auth_info['email'], auth_oid: nil, first_auth_at: nil })
+          .and_return user
+      end
+
+      it 'returns the active user' do
+        expect(authenticate_and_activate!).to be user
+      end
+
+      it "sets the user's name, auth_oid, first_auth_at and last_auth_at" do
```
Comment (willmcb) [inline spec/models/user_spec.rb:115]: Single quotes please :)

**Review by willmcb — APPROVED** [review]
👍 Lets talk it through before merge 👍

File: lib/warden/strategies/azure_ad.rb (hunk 2 above)
Comment (timpeat) [reply lib/warden/strategies/azure_ad.rb:7]: will fix in later pr

**Pull request merged**